airbyte-integrations/connectors/destination-iceberg-v2/src/main/kotlin/io/airbyte/integrations/destination/iceberg/v2/IcebergTypesComparator.kt

```diff
@@ -0,0 +1,231 @@
+/*
+ * Copyright (c) 2024 Airbyte, Inc., all rights reserved.
+ */
+
+package io.airbyte.integrations.destination.iceberg.v2
+
+import jakarta.inject.Singleton
+import org.apache.iceberg.Schema
+import org.apache.iceberg.types.Type
+import org.apache.iceberg.types.Types
+
+/**
+ * Compares two Iceberg [Schema] definitions (including nested structs) to identify:
+ * - New columns that do not exist in the "existing" schema.
+ * - Columns whose data types have changed.
+ * - Columns that no longer exist in the incoming schema (removed).
+ * - Columns that changed from required to optional.
+ */
+@Singleton
+class IcebergTypesComparator {
+
+    companion object {
+        /** Separator used to represent nested field paths: parent~child. */
+        const val PARENT_CHILD_SEPARATOR: Char = '~'
+
+        /**
+         * Returns a fully-qualified field name by appending `child` to `parent` using
+         * [PARENT_CHILD_SEPARATOR]. If `parent` is blank, returns `child` alone.
+         */
+        private fun fullyQualifiedName(parent: String?, child: String): String =
+            if (parent.isNullOrBlank()) child else "$parent$PARENT_CHILD_SEPARATOR$child"
+
+        /**
+         * Splits a fully-qualified name (e.g. `"outer~inner~field"`) into:
+         * ```
+         * parent = "outer~inner"
+         * leaf   = "field"
+         * ```
+         * If there's no [PARENT_CHILD_SEPARATOR], then it's top-level:
+         * ```
+         * parent = ""
+         * leaf   = "outer"
+         * ```
+         */
+        fun splitIntoParentAndLeaf(fqName: String): Pair<String, String> {
+            val idx = fqName.lastIndexOf(PARENT_CHILD_SEPARATOR)
+            return if (idx < 0) {
+                "" to fqName
+            } else {
+                fqName.substring(0, idx) to fqName.substring(idx + 1)
+            }
+        }
+    }
+
+    /**
+     * A data class representing differences between two Iceberg schemas.
+     *
+     * @property newColumns list of fully-qualified column names that are new in the incoming
+     * schema.
+     * @property updatedDataTypes list of fully-qualified column names whose types differ.
+     * @property removedColumns list of fully-qualified column names that are no longer in the
+     * incoming schema.
+     * @property newlyOptionalColumns list of fully-qualified column names that changed from
+     * required -> optional.
+     */
+    data class ColumnDiff(
+        val newColumns: MutableList<String> = mutableListOf(),
```

Comment:
how hard would it be to track `List<List<String>>` instead of encoding the path into a string? then we don't need to do the substring stuff in `splitIntoParentAndLeaf`. Which also makes us more resilient to people putting literal `~` into their field names